eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala

```diff
@@ -0,0 +1,528 @@
+    import f._
+
+    initiateSplice(f, spliceIn_opt = Some(SpliceIn(500_000 sat, pushAmount = 0 msat)))
+    val fundingTx1 = alice.stateData.asInstanceOf[DATA_NORMAL].commitments.latest.localFundingStatus.signedTx_opt.get
+    val watchConfirmed1 = alice2blockchain.expectMsgType[WatchFundingConfirmed]
+    initiateSplice(f, spliceIn_opt = Some(SpliceIn(500_000 sat, pushAmount = 0 msat)))
+    val fundingTx2 = alice.stateData.asInstanceOf[DATA_NORMAL].commitments.latest.localFundingStatus.signedTx_opt.get
+    alice2bob.expectNoMessage()
+    bob2alice.expectNoMessage()
+    alice2blockchain.expectNoMessage()
+    // we now have two unconfirmed splices
+
+    alice ! CMD_FORCECLOSE(ActorRef.noSender)
+    alice2bob.expectMsgType[Error]
+    val aliceCommitTx2 = alice2blockchain.expectMsgType[PublishFinalTx].tx
+    val claimMainDelayed2 = alice2blockchain.expectMsgType[PublishFinalTx].tx
+    val watchConfirmedCommit2 = alice2blockchain.expectMsgType[WatchTxConfirmed]
+    val watchConfirmedClaimMainDelayed2 = alice2blockchain.expectMsgType[WatchTxConfirmed]
+    alice ! WatchFundingSpentTriggered(aliceCommitTx2)
+    alice2blockchain.expectNoMessage()
+
+    // splice 1 confirms
+    watchConfirmed1.replyTo ! WatchFundingConfirmedTriggered(BlockHeight(400000), 42, fundingTx1)
+    alice2bob.forward(bob)
+    alice2blockchain.expectMsgType[WatchFundingSpent]
+    val watchConfirmed2 = alice2blockchain.expectMsgType[WatchFundingConfirmed]
+    alice2blockchain.expectNoMessage()
+
+    // oops! remote commit for splice 1 is published
+    val bobCommitTx1 = bob.stateData.asInstanceOf[PersistentChannelData].commitments.active.find(_.fundingTxIndex == 1).get.localCommit.commitTxAndRemoteSig.commitTx.tx
+    alice ! WatchFundingSpentTriggered(bobCommitTx1)
+    val watchAlternativeConfirmed = alice2blockchain.expectMsgType[WatchAlternativeCommitTxConfirmed]
+    alice2blockchain.expectNoMessage()
+
+    // remote commit tx confirms
+    watchAlternativeConfirmed.replyTo ! WatchAlternativeCommitTxConfirmedTriggered(BlockHeight(400000), 42, bobCommitTx1)
+
+    // we're back to the normal handling of remote commit
+    val claimMain = alice2blockchain.expectMsgType[PublishFinalTx].tx
+    val watchConfirmedRemoteCommit = alice2blockchain.expectMsgType[WatchTxConfirmed]
+    assert(watchConfirmedRemoteCommit.txId == bobCommitTx1.txid)
+    // this one fires immediately, tx is already confirmed
+    watchConfirmedRemoteCommit.replyTo ! WatchTxConfirmedTriggered(BlockHeight(400000), 42, bobCommitTx1)
+    val watchConfirmedClaimMain = alice2blockchain.expectMsgType[WatchTxConfirmed]
+
+    // claim-main tx confirms
+    watchConfirmedClaimMain.replyTo ! WatchTxConfirmedTriggered(BlockHeight(400000), 42, claimMain)
+
+    // done
+    awaitCond(alice.stateName == CLOSED)
+    assert(Helpers.Closing.isClosed(alice.stateData.asInstanceOf[DATA_CLOSING], None).exists(_.isInstanceOf[RemoteClose]))
+  }
+
```

Comment:
I think we need to beef-up the force-close tests, because that's where a bug could cost us funds (which must be avoided at all cost). The following scenarios would be useful to test:

- our peer publishes their commit for an inactive commitment
- our peer publishes a revoked commit for an inactive commitment
- our peer publishes their commit for an active commitment that isn't the latest (not that this is different from your existing test, where the peer publishes their commit *after* we went to the closing state)
- our peer publishes a revoked commit for an active commitment that isn't the latest

We should probably also add tests to verify which watches we set after a restart of eclair while there are active and inactive commitments.

Reply:
The 4 combinations were added in f9b300a9b.

> We should probably also add tests to verify which watches we set after a restart of eclair while there are active and inactive commitments.

I still need to add those (after reconnection and restart).

Reply:
Remaining tests have been added (`put back watches after restart`/`put back watches after restart (inactive)`).